Stop ocaml{c,opt} -c foo.c displaying foo.c on stdout

dra27 opened the pull request:
Microsoft's C Compiler (cl) displays the name of the .c file it is compiling on stdout and there is no switch to suppress this behaviour (tediously).

The assembler (ml{,64}) does the same thing, but fortunately the output of that command can simply be piped to NUL (see [asmcomp/x86_proc.ml, line 243](https://github.com/ocaml/ocaml/blob/trunk/asmcomp/x86_proc.ml#L243)). Obviously that can't be done for invocations of the C compiler, or we'd never see warnings/errors.

The "ideal" solution would be either to pipe stdout and ignore exactly the first line if it is the name of the .c file being compiled or pipe stdout to something equivalent to `tail -n +2` but the former would require merging a large chunk of the pipes and process handling sections of win32unix into ocamlrun (or a similarly invasive change to link the compilers with unix.cm{,x}a) and the latter doesn't have an obvious guaranteed equivalent under Windows.

This solution uses findstr, which is a Windows system component available on all supported versions of Windows.

This code works with .c filenames containing spaces. It would be nice to be able to extend the behaviour to the OCaml build process, but that would involve either converting the invocation of CC to `$(call ...)` (invasive diff) or compiling .c files via the boot compiler (slow), so I'll just have to keep grinding my teeth at all the extra lines in build logs!

dra27 committed:
Filter out the display of .c files by MSVC

Microsoft's C Compiler displays several lines on every invocation - most
of this is removed by specifying the /nologo command line option, but
the compiler still displays the name of every C file it compiles. The
Microsoft Macro Assembler (MASM) does the same thing, but ocamlopt by
default is able to pipe the output of that command directly to NUL, as
the assembler code should never produce errors.

The same cannot be done for invocations of the C compiler, as obviously
syntax errors must be displayed. This relative small cosmetic change
pipes the output of cl to a temporary file and filters out the first
line if it is exactly as expected. The most elegant solution would
require pipes and process handling to be merged from the Unix module
into the compilers.

dra27 commented:
Subsequent testing has revealed a fatal flaw in this approach - the exit status of cl becomes masked by findstr's. I can't see a mechanism for a simple patch in the absence of the Unix module, so desirable as I think having the (std)-output of the compilers being consistent between ports may be, this doesn't seem sanely solvable.

alainfrisch commented:
Would redirecting stdout to a temporary file and parse it from OCaml be possible?

dra27 commented:
Yes - if that's an acceptable solution, I'd be happy to implement that!

alainfrisch commented:
Do you have an idea on the impact on performances?  If it's small enough, I'd be in favor.  Those outputs are really annoying!

dra27 commented:
I would think it should be negligible, relative to what's already going on when compiling? Writing the .obj file must surely eclipse writing lines of status to a file?

dra27 committed:
Use ocamlc as a C compiler in the testsuite

ocamlc and ocamlopt both provide convenient mechanisms to invoke the C
compiler for .c files. Given the filtering out of the .c line for MSVC
now performed by ocamlc and ocamlopt, changing the testsuite to invoke
the C compiler via ocamlc reduces noise from the testsuite _log file.

dra27 commented:
Here we are! I have also added a commit which allows the testsuite to do the same - it means that, apart from a couple of warnings introduced in #396 ([already noted](https://github.com/ocaml/ocaml/pull/396#commitcomment-15298807)), the testsuite now nearly only emits "Running tests " and " ... testing" lines which is very nice. A similar change is needed in flexlink, but I don't know how you want to go about it. I did this:

``` diff
diff --git a/reloc.ml b/reloc.ml
index e009e57..5249d2f 100644
--- a/reloc.ml
+++ b/reloc.ml
@@ -1113,7 +1113,7 @@ let compile_if_needed file =
     let cmd = match !toolchain with
       | `MSVC | `MSVC64 ->
           Printf.sprintf
-            "cl /c /MD /nologo /Fo%s %s %s"
+            "cl /c /MD /nologo /Fo%s %s %s > NUL"
             (Filename.quote tmp_obj)
             (mk_dirs_opt "/I")
             file
```

but it'd probably be better to generalise the error in `run_command` and use that instead?

alainfrisch commented:
Thanks!

As for flexlink, I guess that simply discarding the output is indeed too dangerous.  What about doing about the same as for OCaml?
